File: compare_size.py
#!/usr/bin/env python3
"""Compare compressed and uncompressed sizes of native .so modules across numpy wheel variants."""

import json
import os
import re
import sys
import zipfile
from pathlib import Path

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------
WHEEL_DIR = Path(__file__).parent / "numpy-wheels"
BASELINE = "o2"


def discover_variants():
    return sorted(d.name for d in WHEEL_DIR.iterdir() if d.is_dir())


def find_wheel(variant_dir: Path) -> Path:
    wheels = list(variant_dir.glob("*.whl"))
    if not wheels:
        raise FileNotFoundError(f"No .whl file found in {variant_dir}")
    return wheels[0]


def short_name(filename: str) -> str:
    """numpy/_core/_multiarray_umath.cpython-313-wasm32-emscripten.so → _core/_multiarray_umath"""
    name = filename.removeprefix("numpy/")
    return re.sub(r"\.cpython-[^.]+\.so$", "", name)


def format_bytes(n: int) -> str:
    if n < 1024:
        return f"{n} B"
    if n < 1024 * 1024:
        return f"{n / 1024:.1f} KB"
    return f"{n / (1024 * 1024):.2f} MB"


def diff_str(val: int, base: int) -> str:
    pct = (val / base - 1) * 100
    sign = "+" if pct > 0 else ""
    return f"{format_bytes(val)} ({sign}{pct:.1f}%)"


# ---------------------------------------------------------------------------
# Collect sizes
# ---------------------------------------------------------------------------
def collect_variant(variant: str) -> dict:
    wheel_path = find_wheel(WHEEL_DIR / variant)
    wheel_size = wheel_path.stat().st_size
    so_files = {}

    with zipfile.ZipFile(wheel_path) as zf:
        for info in zf.infolist():
            if info.filename.endswith(".so"):
                so_files[short_name(info.filename)] = {
                    "compressed": info.compress_size,
                    "uncompressed": info.file_size,
                }

    return {"wheel_size": wheel_size, "so_files": so_files}


# ---------------------------------------------------------------------------
# Print tables
# ---------------------------------------------------------------------------
def print_table(all_data: dict, variants: list[str]):
    baseline_data = all_data[BASELINE]
    all_modules = sorted({mod for v in all_data.values() for mod in v["so_files"]})

    name_col = 40

    # --- Wheel size ---
    print("=" * 80)
    print("  WHEEL SIZE COMPARISON")
    print("=" * 80)
    print()

    header = "".ljust(name_col) + "".join(f"| {v:<18}" for v in variants)
    print(header)
    print("-" * len(header))

    row = "Wheel (total)".ljust(name_col)
    for v in variants:
        ws = all_data[v]["wheel_size"]
        if v == BASELINE:
            row += f"| {format_bytes(ws):<18}"
        else:
            row += f"| {diff_str(ws, baseline_data['wheel_size']):<18}"
    print(row)
    print()

    # --- Per-module tables ---
    for label, key in [
        ("NATIVE MODULE SIZES — UNCOMPRESSED", "uncompressed"),
        ("NATIVE MODULE SIZES — COMPRESSED (in wheel)", "compressed"),
    ]:
        print("=" * 80)
        print(f"  {label}")
        print("=" * 80)
        print()

        header = "Module".ljust(name_col) + "".join(f"| {v:<18}" for v in variants)
        print(header)
        print("-" * len(header))

        totals = {v: 0 for v in variants}

        for mod in all_modules:
            row = mod.ljust(name_col)
            for v in variants:
                entry = all_data[v]["so_files"].get(mod)
                base_entry = baseline_data["so_files"].get(mod)
                if not entry:
                    row += f"| {'—':<18}"
                else:
                    totals[v] += entry[key]
                    if not base_entry or v == BASELINE:
                        row += f"| {format_bytes(entry[key]):<18}"
                    else:
                        row += f"| {diff_str(entry[key], base_entry[key]):<18}"
            print(row)

        print("-" * len(header))
        row = "TOTAL".ljust(name_col)
        for v in variants:
            if v == BASELINE:
                row += f"| {format_bytes(totals[v]):<18}"
            else:
                row += f"| {diff_str(totals[v], totals[BASELINE]):<18}"
        print(row)
        print()


# ---------------------------------------------------------------------------
# Markdown output
# ---------------------------------------------------------------------------
def generate_markdown(all_data: dict, variants: list[str]) -> str:
    baseline_data = all_data[BASELINE]
    all_modules = sorted({mod for v in all_data.values() for mod in v["so_files"]})

    md = "## NumPy Native Module Size Comparison\n\n"
    md += f"**Baseline**: `{BASELINE}`\n\n"

    # --- Wheel size ---
    md += "### Wheel Size\n\n"
    md += "| |" + "".join(f" {v} |" for v in variants) + "\n"
    md += "|---|" + "---|" * len(variants) + "\n"
    md += "| Wheel (total) |"
    for v in variants:
        ws = all_data[v]["wheel_size"]
        if v == BASELINE:
            md += f" {format_bytes(ws)} |"
        else:
            md += f" {diff_str(ws, baseline_data['wheel_size'])} |"
    md += "\n\n"

    # --- Per-module tables ---
    for title, key in [
        ("Uncompressed Size", "uncompressed"),
        ("Compressed Size (in wheel)", "compressed"),
    ]:
        md += f"### {title}\n\n"
        md += "| Module |" + "".join(f" {v} |" for v in variants) + "\n"
        md += "|---|" + "---|" * len(variants) + "\n"

        totals = {v: 0 for v in variants}

        for mod in all_modules:
            md += f"| {mod} |"
            for v in variants:
                entry = all_data[v]["so_files"].get(mod)
                base_entry = baseline_data["so_files"].get(mod)
                if not entry:
                    md += " — |"
                else:
                    totals[v] += entry[key]
                    if not base_entry or v == BASELINE:
                        md += f" {format_bytes(entry[key])} |"
                    else:
                        md += f" {diff_str(entry[key], base_entry[key])} |"
            md += "\n"

        md += "| **TOTAL** |"
        for v in variants:
            if v == BASELINE:
                md += f" **{format_bytes(totals[v])}** |"
            else:
                md += f" **{diff_str(totals[v], totals[BASELINE])}** |"
        md += "\n\n"

    return md


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
def main():
    variants = discover_variants()
    print("NumPy Native Module Size Comparison")
    print(f"Variants: {', '.join(variants)}")
    print(f"Baseline: {BASELINE}")
    print()

    all_data = {v: collect_variant(v) for v in variants}

    print_table(all_data, variants)

    # Write markdown
    script_dir = Path(__file__).parent
    md_path = script_dir / "size-results.md"
    md = generate_markdown(all_data, variants)
    md_path.write_text(md)
    print(f"\nMarkdown written to {md_path}")

    # Append to GitHub Step Summary if available
    summary_path = os.environ.get("GITHUB_STEP_SUMMARY")
    if summary_path:
        with open(summary_path, "a") as f:
            f.write(md)
        print("Summary appended to GitHub Step Summary")


if __name__ == "__main__":
    main()

File: compare-size.mjs
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------
const WHEEL_DIR = path.join(__dirname, "numpy-wheels");
const BASELINE = "o2";
const VARIANTS = fs
  .readdirSync(WHEEL_DIR)
  .filter((d) => fs.statSync(path.join(WHEEL_DIR, d)).isDirectory())
  .sort();

// ---------------------------------------------------------------------------
// Minimal zip central-directory parser (metadata only, no extraction)
//
// ZIP end-of-central-directory record (EOCD):
//   offset 0:  signature  0x06054b50
//   offset 16: central directory size (4 bytes)
//   offset 12: central directory offset (4 bytes) — actually offset 16 is size, 12 is start offset
//
// Central directory file header:
//   offset 0:  signature  0x02014b50
//   offset 20: compressed size (4 bytes)
//   offset 24: uncompressed size (4 bytes)
//   offset 28: filename length (2 bytes)
//   offset 30: extra field length (2 bytes)
//   offset 32: file comment length (2 bytes)
//   offset 46: filename (variable)
// ---------------------------------------------------------------------------

function readZipEntries(filePath) {
  const buf = fs.readFileSync(filePath);
  const entries = [];

  // Find EOCD record — scan backwards from end (max 65KB comment)
  let eocdOffset = -1;
  for (let i = buf.length - 22; i >= Math.max(0, buf.length - 65558); i--) {
    if (buf.readUInt32LE(i) === 0x06054b50) {
      eocdOffset = i;
      break;
    }
  }
  if (eocdOffset === -1) throw new Error(`No EOCD record found in ${filePath}`);

  const cdSize = buf.readUInt32LE(eocdOffset + 12);
  const cdOffset = buf.readUInt32LE(eocdOffset + 16);

  let pos = cdOffset;
  while (pos < cdOffset + cdSize) {
    const sig = buf.readUInt32LE(pos);
    if (sig !== 0x02014b50) break;

    const compressedSize = buf.readUInt32LE(pos + 20);
    const uncompressedSize = buf.readUInt32LE(pos + 24);
    const fnLen = buf.readUInt16LE(pos + 28);
    const extraLen = buf.readUInt16LE(pos + 30);
    const commentLen = buf.readUInt16LE(pos + 32);
    const filename = buf.toString("utf8", pos + 46, pos + 46 + fnLen);

    entries.push({ filename, compressedSize, uncompressedSize });
    pos += 46 + fnLen + extraLen + commentLen;
  }

  return entries;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function findWheel(variantDir) {
  const files = fs.readdirSync(variantDir);
  const whl = files.find((f) => f.endsWith(".whl"));
  if (!whl) throw new Error(`No .whl file found in ${variantDir}`);
  return path.join(variantDir, whl);
}

function formatBytes(bytes) {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(2)} MB`;
}

function shortName(filename) {
  // numpy/_core/_multiarray_umath.cpython-313-wasm32-emscripten.so → _core/_multiarray_umath
  return filename
    .replace(/^numpy\//, "")
    .replace(/\.cpython-[^.]+\.so$/, "");
}

// ---------------------------------------------------------------------------
// Collect sizes
// ---------------------------------------------------------------------------

/** @returns {{ soFiles: Map<string, { compressed: number, uncompressed: number }>, wheelSize: number }} */
function collectVariant(variantName) {
  const wheelPath = findWheel(path.join(WHEEL_DIR, variantName));
  const wheelSize = fs.statSync(wheelPath).size;
  const entries = readZipEntries(wheelPath);
  const soFiles = new Map();

  for (const entry of entries) {
    if (entry.filename.endsWith(".so")) {
      soFiles.set(shortName(entry.filename), {
        compressed: entry.compressedSize,
        uncompressed: entry.uncompressedSize,
      });
    }
  }

  return { soFiles, wheelSize };
}

// ---------------------------------------------------------------------------
// Print tables
// ---------------------------------------------------------------------------

function printTable(allData) {
  const baselineData = allData.get(BASELINE);
  if (!baselineData) {
    console.error(`Baseline variant "${BASELINE}" not found`);
    return;
  }

  // Collect all .so module names (union across variants)
  const allModules = new Set();
  for (const { soFiles } of allData.values()) {
    for (const name of soFiles.keys()) allModules.add(name);
  }
  const modules = [...allModules].sort();

  // --- Wheel-level summary ---
  console.log("=".repeat(80));
  console.log("  WHEEL SIZE COMPARISON");
  console.log("=".repeat(80));
  console.log("");

  const nameCol = 40;
  let header = "".padEnd(nameCol);
  for (const v of VARIANTS) header += `| ${v.padEnd(18)}`;
  console.log(header);
  console.log("-".repeat(header.length));

  let row = "Wheel (total)".padEnd(nameCol);
  for (const v of VARIANTS) {
    const { wheelSize } = allData.get(v);
    const baseSize = baselineData.wheelSize;
    const diff = ((wheelSize / baseSize - 1) * 100).toFixed(1);
    const label =
      v === BASELINE
        ? formatBytes(wheelSize)
        : `${formatBytes(wheelSize)} (${diff > 0 ? "+" : ""}${diff}%)`;
    row += `| ${label.padEnd(18)}`;
  }
  console.log(row);
  console.log("");

  // --- Uncompressed .so sizes ---
  console.log("=".repeat(80));
  console.log("  NATIVE MODULE SIZES — UNCOMPRESSED");
  console.log("=".repeat(80));
  console.log("");

  header = "Module".padEnd(nameCol);
  for (const v of VARIANTS) header += `| ${v.padEnd(18)}`;
  console.log(header);
  console.log("-".repeat(header.length));

  let totalUncompressed = Object.fromEntries(VARIANTS.map((v) => [v, 0]));

  for (const mod of modules) {
    let row = mod.padEnd(nameCol);
    for (const v of VARIANTS) {
      const entry = allData.get(v).soFiles.get(mod);
      const baseEntry = baselineData.soFiles.get(mod);
      if (!entry) {
        row += `| ${"—".padEnd(18)}`;
      } else {
        totalUncompressed[v] += entry.uncompressed;
        if (!baseEntry || v === BASELINE) {
          row += `| ${formatBytes(entry.uncompressed).padEnd(18)}`;
        } else {
          const diff = ((entry.uncompressed / baseEntry.uncompressed - 1) * 100).toFixed(1);
          row += `| ${(formatBytes(entry.uncompressed) + ` (${diff > 0 ? "+" : ""}${diff}%)`).padEnd(18)}`;
        }
      }
    }
    console.log(row);
  }

  console.log("-".repeat(header.length));
  row = "TOTAL".padEnd(nameCol);
  for (const v of VARIANTS) {
    const base = totalUncompressed[BASELINE];
    const val = totalUncompressed[v];
    const diff = ((val / base - 1) * 100).toFixed(1);
    const label =
      v === BASELINE
        ? formatBytes(val)
        : `${formatBytes(val)} (${diff > 0 ? "+" : ""}${diff}%)`;
    row += `| ${label.padEnd(18)}`;
  }
  console.log(row);
  console.log("");

  // --- Compressed .so sizes ---
  console.log("=".repeat(80));
  console.log("  NATIVE MODULE SIZES — COMPRESSED (in wheel)");
  console.log("=".repeat(80));
  console.log("");

  header = "Module".padEnd(nameCol);
  for (const v of VARIANTS) header += `| ${v.padEnd(18)}`;
  console.log(header);
  console.log("-".repeat(header.length));

  let totalCompressed = Object.fromEntries(VARIANTS.map((v) => [v, 0]));

  for (const mod of modules) {
    let row = mod.padEnd(nameCol);
    for (const v of VARIANTS) {
      const entry = allData.get(v).soFiles.get(mod);
      const baseEntry = baselineData.soFiles.get(mod);
      if (!entry) {
        row += `| ${"—".padEnd(18)}`;
      } else {
        totalCompressed[v] += entry.compressed;
        if (!baseEntry || v === BASELINE) {
          row += `| ${formatBytes(entry.compressed).padEnd(18)}`;
        } else {
          const diff = ((entry.compressed / baseEntry.compressed - 1) * 100).toFixed(1);
          row += `| ${(formatBytes(entry.compressed) + ` (${diff > 0 ? "+" : ""}${diff}%)`).padEnd(18)}`;
        }
      }
    }
    console.log(row);
  }

  console.log("-".repeat(header.length));
  row = "TOTAL".padEnd(nameCol);
  for (const v of VARIANTS) {
    const base = totalCompressed[BASELINE];
    const val = totalCompressed[v];
    const diff = ((val / base - 1) * 100).toFixed(1);
    const label =
      v === BASELINE
        ? formatBytes(val)
        : `${formatBytes(val)} (${diff > 0 ? "+" : ""}${diff}%)`;
    row += `| ${label.padEnd(18)}`;
  }
  console.log(row);
}

// ---------------------------------------------------------------------------
// Markdown output
// ---------------------------------------------------------------------------

function generateMarkdown(allData) {
  const baselineData = allData.get(BASELINE);
  if (!baselineData) return "";

  const allModules = new Set();
  for (const { soFiles } of allData.values()) {
    for (const name of soFiles.keys()) allModules.add(name);
  }
  const modules = [...allModules].sort();

  let md = "## NumPy Native Module Size Comparison\n\n";
  md += `**Baseline**: \`${BASELINE}\`\n\n`;

  // --- Wheel size ---
  md += "### Wheel Size\n\n";
  md += "| |";
  for (const v of VARIANTS) md += ` ${v} |`;
  md += "\n|---|";
  for (const _v of VARIANTS) md += "---|";
  md += "\n| Wheel (total) |";
  for (const v of VARIANTS) {
    const { wheelSize } = allData.get(v);
    const baseSize = baselineData.wheelSize;
    if (v === BASELINE) {
      md += ` ${formatBytes(wheelSize)} |`;
    } else {
      const diff = ((wheelSize / baseSize - 1) * 100).toFixed(1);
      md += ` ${formatBytes(wheelSize)} (${diff > 0 ? "+" : ""}${diff}%) |`;
    }
  }
  md += "\n\n";

  // Helper to build a size table
  function sizeTable(title, sizeKey) {
    md += `### ${title}\n\n`;
    md += "| Module |";
    for (const v of VARIANTS) md += ` ${v} |`;
    md += "\n|---|";
    for (const _v of VARIANTS) md += "---|";
    md += "\n";

    const totals = Object.fromEntries(VARIANTS.map((v) => [v, 0]));

    for (const mod of modules) {
      md += `| ${mod} |`;
      for (const v of VARIANTS) {
        const entry = allData.get(v).soFiles.get(mod);
        const baseEntry = baselineData.soFiles.get(mod);
        if (!entry) {
          md += " — |";
        } else {
          totals[v] += entry[sizeKey];
          if (!baseEntry || v === BASELINE) {
            md += ` ${formatBytes(entry[sizeKey])} |`;
          } else {
            const diff = ((entry[sizeKey] / baseEntry[sizeKey] - 1) * 100).toFixed(1);
            md += ` ${formatBytes(entry[sizeKey])} (${diff > 0 ? "+" : ""}${diff}%) |`;
          }
        }
      }
      md += "\n";
    }

    md += `| **TOTAL** |`;
    for (const v of VARIANTS) {
      const base = totals[BASELINE];
      const val = totals[v];
      if (v === BASELINE) {
        md += ` **${formatBytes(val)}** |`;
      } else {
        const diff = ((val / base - 1) * 100).toFixed(1);
        md += ` **${formatBytes(val)} (${diff > 0 ? "+" : ""}${diff}%)** |`;
      }
    }
    md += "\n\n";
  }

  sizeTable("Uncompressed Size", "uncompressed");
  sizeTable("Compressed Size (in wheel)", "compressed");

  return md;
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

console.log("NumPy Native Module Size Comparison");
console.log(`Variants: ${VARIANTS.join(", ")}`);
console.log(`Baseline: ${BASELINE}`);
console.log("");

const allData = new Map();
for (const v of VARIANTS) {
  allData.set(v, collectVariant(v));
}

printTable(allData);

// Write markdown
const mdPath = path.join(__dirname, "size-results.md");
const md = generateMarkdown(allData);
fs.writeFileSync(mdPath, md);
console.log(`\nMarkdown written to ${mdPath}`);

// Append to GitHub Step Summary if available
if (process.env.GITHUB_STEP_SUMMARY) {
  fs.appendFileSync(process.env.GITHUB_STEP_SUMMARY, md);
  console.log("Summary appended to GitHub Step Summary");
}
